include/ipc/ipc_service.h

```diff
@@ -170,6 +170,80 @@ int ipc_service_register_endpoint(const struct device *instance,
  */
 int ipc_service_send(struct ipc_ept *ept, const void *data, size_t len);
 
+/** @brief Get an empty TX buffer to be sent using @ref ipc_service_send_nocopy
+ *
+ *  This function can be called to get an empty TX buffer so that the
+ *  application can directly put its data into the sending buffer without copy
+ *  from an application buffer.
+ *
+ *  It is the application responsibility to correctly fill the allocated TX
+ *  buffer with data and passing correct parameters to @ref
+ *  ipc_service_send_nocopy function to perform data no-copy-send mechanism.
+ *
+ *  @param ept Registered endpoint by @ref ipc_service_register_endpoint.
+ *  @param data Pointer to the empty TX buffer.
+ *  @param len Pointer to store the TX buffer size.
+ *  @param wait Boolean, wait or not for buffer to become available
+ *
+ *  @retval -EIO when no backend is registered or send hook is missing from
+ *	    backend.
+ *  @retval -EINVAL when instance or endpoint is invalid.
+ *  @retval -ENOTSUP when this is not supported by backend.
+ *  @retval -ENOBUFS when there are no TX buffers available.
+ *
+ *  @retval 0 on success.
+ *  @retval other errno codes depending on the implementation of the backend.
+ */
+int ipc_service_get_buffer(struct ipc_ept *ept, void **data, uint32_t *len, int wait);
+
+/** @brief Send data in a TX buffer reserved by @ref ipc_service_get_buffer
+ *         using the given IPC endpoint.
+ *
+ *  This is equivalent to @ref ipc_service_send but in this case the TX buffer
+ *  has been obtained by using @ref ipc_service_get_buffer.
+ *
+ *  The application has to take the responsibility for getting the TX buffer
+ *  using @ref ipc_service_get_buffer and filling the TX buffer with the data.
+ *
+ *  After the @ref ipc_service_send_nocopy function is issued the TX buffer is
+ *  no more owned by the sending task and must not be touched anymore unless
+ *  the function fails and returns an error.
```

Comment:
It looks a function to release a TX buffer is needed. If a user calls `ipc_service_send_nocopy()` which returns one of the error responses, user may choose to discard the buffer.

Reply:
Right, 
A related issue as just be entered in OpenAMP:
https://github.com/OpenAMP/open-amp/issues/361

Reply:
I would not add a function but assume the backend will always take ownership regardless if send is successful or not.

The only reason to keep it separated would be if the user was responsible for send retry. But we need retry only if buffer is not available which is not the case here -we already grabbed a buffer earlier. I don't see a point in complicating the API.

Reply:
We need a separate `send()` function for this because the backend must be able to differentiate the nocopy case vs the regular case. See for example OpenAMP functions `rpmsg_send_offchannel_nocopy()` vs `rpmsg_send_offchannel()`.